Work toward supporting code generation for models with imports

nickerso opened the pull request:
Addressing #386.

nickerso committed:
Adding initial test to demonstrate expected behaviour when using a model with imports.

Adding a new test resource with a model with imports which is the same model as the algebraicEqnComputedVarOnRhs
generator test but with an import. I (currently) expect this to result in the same code as the non-import version.

nickerso committed:
starting to think about flattening prior to code generation

hsorby commented on src/api/libcellml/model.h:
```diff
@@ -251,6 +251,16 @@ class LIBCELLML_EXPORT Model: public ComponentEntity, public std::enable_shared_
+	/**
+	 * @brief Flatten all imports in this model.
+	 *
+	 * Flatten this model to remove all imports. If any unresolved imports are found
```
I think flatten should be changed to actually describe what is going on.  Flatten here has too much of a contextual meaning.

agarny commented on tests/generator/generator.cpp:
```diff
@@ -393,6 +393,8 @@ TEST(Generator, algebraicEqnComputedVarOnRhsImport)
 	model->resolveImports(resourcePath("generator/imports/algebraic_eqn_computed_var_on_rhs/"));
 	EXPECT_FALSE(model->hasUnresolvedImports());
 
+	model->flatten();
```
To have a method to "flatten" a model is fine by me, but... when it comes to code generation, I would expect the generator to realise that the model contains some imports. Therefore, I would expect the generator to create a "flattened" version of that model before generating code for it.

nickerso replied:
my current thinking is that flatten would be changing the current model - I would not expect code generation to change the model. Hence not calling the flatten within the code generation. Code generation should generate code for the model it is given (or fail), not make a new model and generate code for that.

An alternative is to have flatten create a new model which is the flattened version of the source model, but I still don't think that is something that the generator should be doing either. A flattened model can be used for things outside of generation, or at least the same flattened model could be used with multiple profiles...

hsorby replied:
I think that this method should return the flattened model and not flatten the model that it is.

nickerso replied:
Talking it over with @agarny and @hsorby, it is clear that the flatten method should create a new model. I still think the generator should be working on the model I give it, not making its own flattened model internally and generating code for that.

hsorby replied:
I think so too, returning a flattened model from `model->flatten` would work in nicely with this.

agarny replied:
+1 on the flatten method creating a new model. However, if we are to do this, then we might want to rename the method to something like `model->flattenedModel()`? `flatten()` on its own tells me that `model` is going to be flattened, which is not what we want.

As for the code generator working on the model it is given... I guess I could eventually get convinced that it's the way to go...

nickerso committed:
rethink model flatten method and improve docs.

nickerso committed:
Merge branch 'develop' of https://github.com/cellml/libcellml into code-gen-imports-i386

nickerso committed:
hacking things for model flattening, but getting confused.

Need to explore copying models and components independent of the model flattening.

nickerso committed:
tweak variable names to help debug model flattening

nickerso committed:
basic single level component import flattening kind of working.

nickerso committed:
renaming components and variables such that the generated code should match non-import version

nickerso committed:
adding initial test for a more complex import test using HH model

nickerso committed:
adding initial test for a more complex import test using HH model

nickerso committed:
fixing a typo in the test HH with imports models

nickerso committed:
move potassium channel to an imported model

nickerso committed:
traverse children components when flattening a model

nickerso committed:
adding models from VPH-MIP version of HH where a single HH gating variable component is reused for each of the gating variables.

Adding a new generator test to try and generate code for this model, which is next level complex

nickerso committed:
trying to make sure component names are unique

nickerso committed:
Merging prime develop and resolving conflicts.

Still more to do to get this building with the large changes that have gone into
prime.

nickerso committed:
Rethinking model flattening API.

With the move to create methods, starting to make more sense that model flattening
is a utility method with a wrapper static method in the Model class. This is the first
change toward that, but more to do.

nickerso committed:
Merge branch 'develop' of https://github.com/cellml/libcellml into code-gen-imports-i386

nickerso committed:
Getting the code to build, not yet doing anything useful.

nickerso committed:
basic flattening working again, initial test not passing yet

nickerso committed:
Merge branch 'develop' of https://github.com/cellml/libcellml into code-gen-imports-i386

nickerso committed:
thinking about tracking component aliases through the import hierarchy.

This idea is a bit of a dead end, but wanting to archive these changes in case
I come back to this approach.

nickerso committed:
Merge branch 'develop' of https://github.com/cellml/libcellml into code-gen-imports-i386

nickerso committed:
Merge branch 'develop' of https://github.com/cellml/libcellml into code-gen-imports-i386

nickerso committed:
Starting to work on model flattening tests.

Separate to the code generation issue, need to better test model flattening with models
that may not be suitable for code generation. This added test also raises an issue with
model printing.

nickerso committed:
Merge in prime/develop, resolve conflicts, and update to match recent changes in the create methods.

nickerso committed:
Adding initial variable clones and thinking about tracking hierarchical imports

nickerso committed:
Merge branch 'develop' of https://github.com/cellml/libcellml into code-gen-imports-i386

nickerso committed:
starting to add variable equivalences

nickerso commented:
closing in favour of #483.
